Improve trigonometric function evaluation and logarithmic rewrite

ylemkimon opened the pull request:
- Added evaluation of sin, cos, tan, cot for asec, acsc args, i.e. `cos(asec(x)`
- Fixed logarithmic rewrites of atan, acot
- Added base class for inverse hyperbolic functions

ylemkimon committed:
Add eval of TrigonometricFunc for asec, acsc args

Added evaluation of TrigonometricFunctions - sin, cos, tan, cot - for
asec, acsc args.
Reference :
https://en.wikipedia.org/wiki/Inverse_trigonometric_functions

ylemkimon committed:
Fix incorrect _eval_rewrite_as_log for atan, acot

Logarithmic rewrite of inverse hyperbolic secant is incorrect(complex
logarithm)
Reference :
https://en.wikipedia.org/wiki/Inverse_trigonometric_functions

ylemkimon committed:
Add/Fix test cases for TrigonometricFunction

Added test cases of TrigonometricFunction - sin, cos, tan, cot -
evaluation for acsc, asec args and fixed test cases of logarithmic
rewrite for atan and acot

ylemkimon committed:
Add InverseHyperbolicFunction base class

Added base class for inverse hyperbolic functions(asinh, acosh, atanh,
acsch, asech, acoth) - InverseHyperbolicFunction.

ylemkimon committed:
Add argument test for InverseHyperbolicFunction

Added
test_sympy__functions__elementary__hyperbolic__InverseHyperbolicFunction,
fixes AssertionError: Missing classes:
sympy.functions.elementary.hyperbolic.InverseHyperbolicFunction.

smichr commented on sympy/functions/elementary/hyperbolic.py:
```diff
@@ -771,7 +771,13 @@ def _sage_(self):
+class InverseHyperbolicFunction(Function):
+    """Base class for inverse hyperbolic functions."""
+
+    pass
```
Could you explain why this is necessary?

ylemkimon replied:
@smichr This was for #13110 .
Since it is closed, it is unnecessary for now, but I put and left it here for future usage in simplification, manipulation, etc.

ylemkimon replied:
InverseTrigonometricFunction in trigonometric.py also seems to have no usage.

smichr commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -2240,8 +2272,8 @@ def _eval_is_real(self):
-        return S.ImaginaryUnit/2 * (log(
-            (S(1) - S.ImaginaryUnit * x)/(S(1) + S.ImaginaryUnit * x)))
+        return S.ImaginaryUnit/2 * (log(S(1) - S.ImaginaryUnit * x)
+            - log(S(1) + S.ImaginaryUnit * x))
```
Was there a problem with the previous form?

ylemkimon replied:
@smichr Because it's complex logarithm, two forms can differ by multiples of pi.
i.e., `assert atan(5*I).evalf() == atan(x).rewrite(log).evalf(subs={x:5*I})` fails with previous form.

smichr commented:
Thanks for the clarification. This looks good, then. Last question: have you checked the log-forms numerically? I see that you test them symbolically, but it wouldn't hurt to include a check -- not necessarily in the suite, just a manual "pre-launch" test.

ylemkimon committed:
Fix incorrect _eval_rewrite_as_log for acot

Logarithmic rewrite of inverse cotangent is incorrect(complex
logarithm), fixed _eval_rewrite_as_log and its test case.
Reference :
https://en.wikipedia.org/wiki/Inverse_trigonometric_functions

ylemkimon commented:
@smichr It turns out that the new form of logarithmic rewrite of atan is incorrect too. (I should've double checked. I'm sorry.) Fixed it and added new commit.

With new commit, following test seems to run successfully.

```python
>>> from sympy import *
>>> x = Symbol('x')
>>> from sympy.utilities.randtest import verify_numerically
>>> for i in range(1, 1000000):
...     assert verify_numerically(atan(x), atan(x).rewrite(log), z = x, tol = 1.0e-15, a = -10, b = -10, c = 10, d = 10)
...     assert verify_numerically(acot(x), acot(x).rewrite(log), z = x, tol = 1.0e-15, a = -10, b = -10, c = 10, d = 10)
...     print i, "/1000000"
```

smichr commented:
Hmm, 1-1/x is (x-1)/x, and similar for the other term. The love expansion rules are not fresh in my mind--can't the ratio be rewritten as log(1-x)-log(x) and then the log terms cancel? Or is this a case where certain values of x give a result off by some multiple?

jksuom commented:
There does not seem to exist universally accepted rules for choosing the values of inverse trigonometric functions on a branch cut. One [standard](http://dlmf.nist.gov/4.23) only says: "Care needs to be taken on the cuts". `mpmath` has fixed some choices. We could copy them, but I don't think that different choices would necessarily be impossible if that would lead to simpler expressions.

ylemkimon commented:
@smichr Yes, since it's complex logarithm, depending on values of x-I and x, two forms can differ by multiples of pi.

@jksuom I think it should be consistent with evalf and relation to other functions(_eval_rewrite_as_~).

smichr commented:
I'm ok with this if you are, @jksuom.

jksuom commented:
This apparently means that the conventions of `mpmath` are followed, which is ok by me. Perhaps there could be some numerical tests to confirm this.

ylemkimon commented:
@jksuom As commented above, following test seems to run successfully.

```python
>>> from sympy import *
>>> x = Symbol('x')
>>> from sympy.utilities.randtest import verify_numerically
>>> for i in range(1, 1000000):
...     assert verify_numerically(atan(x), atan(x).rewrite(log), z = x, tol = 1.0e-15, a = -10, b = -10, c = 10, d = 10)
...     assert verify_numerically(acot(x), acot(x).rewrite(log), z = x, tol = 1.0e-15, a = -10, b = -10, c = 10, d = 10)
...     print i, "/1000000"
```

jksuom commented:
I think that a couple of test points on different parts of the branch cuts would suffice to verify that the correct boundary value has been chosen. Those are the critical points where different values may occur. Outside the branch cuts all given expressions take the principal value. (The branch cuts of `atan` and `acot` lie on the imaginary axis.)

ylemkimon commented:
@jksuom As you said, since branch cut for atan and acot is (-i∞, -i]∪[i, i∞) and (-i, i), respectively, above test with a = 0 and c = 0 would do it, and it runs successfully too.

jksuom commented:
> above test with a = 0 and c = 0 would do it, and it runs successfully too.

I have no doubts about that. But please add a few relevant unit tests of the type

`assert atan(5*I).evalf() == atan(x).rewrite(log).evalf(subs={x:5*I})`.

ylemkimon committed:
Add test cases for logarithmic rewrite of atan, acot

Added test cases for logarithmic rewrite evaluation of atan, acot to
verify that the correct boundary value has been chosen.

ylemkimon commented:
@jksuom Ah, I though you meant manual pre-launch test as @smichr mentioned above. Added test cases.

jksuom commented on sympy/functions/elementary/tests/test_trigonometric.py:
```diff
@@ -857,6 +857,8 @@ def test_atan_rewrite():
     assert atan(x).rewrite(asec) == sqrt(x**2)*asec(sqrt(x**2 + 1))/x
     assert atan(x).rewrite(acsc) == (-acsc(sqrt(x**2 + 1)) + pi/2)*sqrt(x**2)/x
 
+    assert atan(5*I).evalf() == atan(x).rewrite(log).evalf(subs={x:5*I})
```
Thanks, this should take care of the cut from `I` to `oo*I`. To fully cover both pieces of the cut another point on the negative imaginary axis is needed. Similarly, the branch cut of `acot` consists of two pieces separated by the singularity at the origin.

ylemkimon replied:
@jksuom Added 😄

ylemkimon committed:
Add more test cases for logarithmic rewrite of atan, acot

Added test cases for the other side of branch cuts for logarithmic
rewrite evaluation of atan, acot to verify that the correct boundary
value has been chosen.

jksuom commented:
Thanks, I think this is ready to be merged.
